R/chat-structured.R

```diff
@@ -84,3 +68,33 @@ convert_from_type <- function(x, type) {
     x
   }
 }
+
+list_to_atomic <- function(x, type) {
+  r_type <- switch(
+    type,
+    boolean = "logical",
+    integer = "integer",
+    number = "double",
+    string = "character",
+    cli::cli_abort("Unknown type {type}", .internal = TRUE)
+  )
+
+  if (length(x) == 0) {
+    return(vector(r_type, 0))
+  }
+
+  types <- map_chr(x, typeof)
+  if (r_type == "integer") {
+    x[types == "double"] <- lapply(x[types == "double"], as.integer)
+    types[types == "double"] <- "integer"
+  } else if (r_type == "double") {
+    x[types == "integer"] <- lapply(x[types == "integer"], as.double)
+    types[types == "integer"] <- "double"
+  }
```

Comment:
```suggestion
  if (r_type == "integer") {
    doubles <- which(types == "double")
    x[doubles] <- lapply(x[doubles], as.integer)
    types[doubles] <- "integer"
  } else if (r_type == "double") {
    integers <- which(types == "integer")
    x[integers] <- lapply(x[integers], as.double)
    types[integers] <- "double"
  }
```

Depending on how much you care about performance